### PR TITLE
feat: add Pi evidence provider (#244)

### LINKED ISSUE #244: Promote Pi to a dedicated bounded full-evidence provider
## Summary

Promote Pi from the generic JSONL fallback to a dedicated, bounded full-evidence provider, using the Codex provider in #243 as the implementation template.

Pi discovery already exists and Pi-shaped metrics parse through shared JSONL heuristics. The missing work is an explicit Pi format contract, evidence streaming, source-validity handling, capabilities, persistence/backfill, UI/report coverage, and a privacy-safe characterization suite.

This issue should be implementation-ready for a later `use-workflow` kickoff. It does **not** include implementation.

## Dependencies and sequencing

- Implement after #243 merges. Reuse its adapter, routing, capability, characterization, and backfill patterns rather than recreating those seams.
- Resolve or explicitly adopt the current outcome of #229 for unknown record/content types before finalizing Pi coverage semantics.
- Apply the badge/report coverage behavior established by #242 and the final #229 policy. Do not invent a Pi-only interpretation of `Complete`, `Partial`, or unrecognized types.

### Discovery is already implemented

`crates/antiburn-local/src/discovery/agents/pi.rs` contains `PiExplorer`:

- scans `~/.pi/agent/sessions/**` for `.jsonl` sessions;
- reads `cwd` from the first `type: "session"` row;
- recovers the session ID from `{ISO-timestamp}_{UUIDv7}.jsonl` filenames;
- classifies Pi as a CLI source;
- supplies file-backed `SessionSource` values to the existing pipeline.

`AgentKind::Pi` and the `pi` presentation entry already exist.

### Analysis is still generic

`crates/antiburn-local/src/analysis/vendors/mod.rs` currently sends `pi` to `GenericJsonlAdapter`. The registry test explicitly keeps Pi in `generic_fallback_vendors_have_no_dedicated_adapter`.

The shared parser in `analysis/vendors/jsonl.rs` already recognizes useful parts of Pi records, including:

- `message.role`;
- `text`, `thinking`, and `toolCall` content blocks;
- `toolResult` messages and `isError`;
- `input`, `output`, `cacheRead`, and `cacheWrite` usage fields.

`crates/antiburn-local/src/analysis/tests.rs` proves Pi-shaped tool classification and token accounting, but these tests do not establish a complete Pi schema or evidence contract.

### Pi is not in the evidence cohort

- `has_dedicated_adapter("pi")` is false.
- `apps/desktop/src/lib/presentation/agents.ts` sets `AGENTS.pi.supportsAnalysis` to false.
- The desktop evidence routing/capability map and evidence cohort do not include Pi.
- Existing Pi sessions do not receive the dedicated evidence pass/backfill that #243 introduces for Codex.

## Problem

Generic fallback behavior is intentionally best effort. It cannot support an honest agent-level evidence claim because it does not define:

- which Pi row and content-block types are structurally recognized;
- how malformed, truncated, oversized, or unknown records affect coverage;
- which evidence capabilities Pi can provide;
- bounded append-only reads and source mutation outcomes;
- Pi-specific session/model/compaction/fork semantics;
- characterization against real Pi behavior;
- migration and reprocessing of existing Pi sessions;
- synchronized engine, DTO, report, badge, and frontend support gating.

As a result, useful metrics may appear accidentally while the product correctly reports that Pi lacks dedicated analysis support.

## Desired outcome

Pi becomes a first-class provider with the same safety properties as the Codex adapter from #243:

1. a dedicated adapter and explicit Pi schema recognition;
2. one-pass bounded metrics and evidence streaming;
3. claimed source validation for live append-only files;
4. an accurate `SourceCapabilities::pi()` declaration;
5. desktop routing, persistence, cohort membership, and backfill;
6. privacy-safe characterization fixtures and goldens;
7. synchronized UI/report support gating;
8. no raw transcript content in evidence, persistence, DTOs, logs, fixtures, or issue/PR discussion.

Do not flip `supportsAnalysis` or describe Pi as supported until the complete path above is in place.

## Pi format to characterize

Use current Pi v3 sessions as the initial format target. Confirm every assumption during implementation against the installed Pi version and privacy-safe local inspection.

Expected record families include:

- `session` header records, including version, session identity, timestamp, cwd, and optional parent-session metadata;
- `message` records with top-level identity, parent linkage, timestamp, and nested message data;
- user, assistant, and tool-result roles;
- `text`, `thinking`, and `toolCall` content blocks;
- tool results with `toolCallId`, `toolName`, content, and `isError`;
- assistant model/provider/API metadata where present;
- usage fields `input`, `output`, `cacheRead`, `cacheWrite`, and `totalTokens`;
- model/thinking configuration changes;
- compaction/session-information/custom rows;
- unknown future row and content-block types.

The implementation must verify rather than assume:

- which timestamp is authoritative for turn ordering and duration;
- whether usage is per-turn, cumulative, or mixed for every observed shape;
- whether replayed/inherited records occur and how ownership is established;
- whether `parentSession`, row parent IDs, or other fields represent a fork, an ordinary message tree, or both;
- whether compaction records change context/accounting semantics;
- which metadata is safe and useful to retain under the existing normalized/evidence model.

### 1. Dedicated adapter

Add a Pi adapter under `crates/antiburn-local/src/analysis/vendors/` and route `adapter_for("pi")` to it.

- Preserve the invariant that every unknown vendor still resolves to the generic fallback.
- Keep `has_dedicated_adapter` derived from `adapter_for`; do not introduce another backend vendor list.
- Move Pi from the generic-fallback registry test to the dedicated-adapter test.
- Reuse shared JSONL helpers only where their semantics are proven to match Pi. Keep Pi-specific recognition and state in the Pi adapter.

### 2. Bounded record streaming

Follow the post-#243 Codex implementation:

- implement `visit` and `visit_claimed`;
- use `BoundedJsonlReader`, existing record-size limits, cancellation, `SourceClaim`, and `PinnedSource`;
- return the correct `VisitOutcome` for a full accepted file, an accepted pinned prefix, or a changed source;
- never materialize an unbounded transcript or retain transcript text after parsing;
- skip/degrade unusable records according to the shared coverage policy rather than failing the whole session.

Test source replacement, truncation, append-after-claim, short reads, and cancellation where the shared harness supports them.

### 3. Metrics parity

The dedicated adapter must preserve or improve current Pi metrics:

- roles and timestamps;
- text length and thinking detection;
- tool calls, tool categories, test-command classification, and tool errors;
- model attribution;
- four disjoint usage buckets: input, output, cache read, and cache creation;
- context-window and cache-write availability only when the source can prove them.

Add parity tests showing that promotion does not regress the Pi metrics already pinned in `analysis/tests.rs`.

### 4. Evidence capability matrix

Add `SourceCapabilities::pi()` after inspecting the post-#243 capability model. For every capability, document why it is true or false and pin it with tests.

At minimum, decide and characterize:

- request-context tokens;
- cache-write tokens;
- timestamps and ordering;
- tool invocations;
- skill/MCP attribution;
- tool definitions;
- model identity and model transitions;
- token classes;
- reasoning-effort/thinking tiers;
- fast/service tiers;
- subagent relationships and child models;
- compaction boundaries;
- thread identity;
- quota incidents;
- harness version.

A capability must be `true` only when Pi records and adapter output fully support it. Unsupported is preferable to inferred or partial support that the evidence model cannot represent honestly.

### 5. Row and content recognition

Implement structural recognition for Pi rows and nested content blocks.

- Known semantic rows emit normalized metrics/evidence.
- Known inert housekeeping rows are treated according to the final #229 policy.
- Unknown row types, unknown `customType` values, and unknown content-block types are surfaced and affect coverage exactly as #229 specifies.
- Do not silently report clean/complete coverage when an unrecognized structure could carry relevant evidence.
- Bound discriminator strings and counts using the existing evidence limits.

If #229 requires DTO or report changes for `unrecognizedTypes`, include Pi in those shared changes rather than creating a Pi-only field.

### 6. Live-file failure behavior

Characterize and test:

- a truncated final line while Pi is still writing;
- malformed JSON in the middle of a file;
- oversized records;
- missing required fields;
- unsupported schema versions;
- an empty or header-only session;
- a file that changes identity or shrinks during analysis.

These conditions must produce the shared, honest coverage/outcome state and must not crash or discard all usable preceding records unless source validity requires rejecting the pass.

### 7. Model and configuration changes

Track Pi model identity from message metadata and explicit model-change records where present.

- Preserve transitions rather than assigning the final model to all earlier turns.
- Treat provider/API/account-like identifiers as input to normalization only when necessary; do not persist opaque account identity or credentials through this issue.
- Decide explicitly whether thinking-level changes map to an existing effort-tier capability. Do not add a new evidence shape without the corresponding schema/revision review.
- Explicitly ignore `usage.cost` unless an existing normalized contract requires it. Pricing/cost expansion is not part of this issue.

### 8. Fork, inherited-history, and subagent behavior

Pi fixtures in Cadence include parent/fork cases, but Antiburn must independently prove the applicable semantics.

- Distinguish a real fork from ordinary `parentId` message threading, compaction, and unrelated sessions.
- Determine whether a child transcript contains inherited parent rows and, if so, identify a deterministic ownership boundary.
- Do not double-count inherited usage or evidence in child ownership.
- Do not infer subagent relationships solely from a generic message tree.
- Emit fork/subagent evidence only if the post-#243 model can represent it completely. Otherwise mark the capability unsupported and document the follow-up.

### 9. Desktop routing and persistence

On the post-#243 baseline:

- add Pi to the vendor capability routing in `apps/desktop/src-tauri/src/analysis.rs`;
- ensure the evidence worker routes Pi through the streaming path and never defaults unknown vendors to Claude;
- add Pi to `evidence_cohort()` in `apps/desktop/src-tauri/src/agents.rs`;
- add the next store migration to enqueue/backfill existing Pi sessions, mirroring the Codex migration;
- update cohort, queue, stale-revision, source-missing, and persistence tests as applicable;
- preserve atomic publication of metrics and evidence for one source generation.

Do not hard-code a migration number in the implementation plan until #243 has merged and the current schema version is known.

### 10. Revisions

Review and state the reason for every revision decision:

- bump `PARSER_REVISION` because Pi parsing/recognition changes;
- bump `ANALYZER_REVISION` only if normalized analysis semantics change;
- bump `EVIDENCE_SCHEMA_REVISION` only if the persisted evidence contract changes, including any shared #229 additions;
- leave `METRICS_SCHEMA_REVISION` unchanged unless the metrics shape changes.

Tests must prove existing Pi sessions become eligible for reprocessing/backfill under the selected revisions.

### 11. UI, badges, and reports

Only after the provider path is complete:

- set `AGENTS.pi.supportsAnalysis` to true in `apps/desktop/src/lib/presentation/agents.ts`;
- update frontend and backend mirror tests;
- verify DTO support state agrees with `has_dedicated_adapter("pi")`;
- verify Pi evidence appears in the pane/report without a generic unsupported empty state;
- verify partial/unknown coverage affects badges and reports according to #242 and #229;
- update documentation that enumerates full-evidence providers.

No new UI design or styling is expected.

## Privacy and fixture policy — binding

Real Pi session files under `~/.pi` are available for local reference, but they are private source material.

They may be inspected locally only to identify schema, invariants, and edge cases. They must **never** be:

- committed to this repository or any branch;
- attached to an issue or pull request;
- pasted or quoted verbatim in prompts, comments, plans, logs, test output, or reports;
- copied into temporary artifacts that leave the local machine;
- represented by hashes or provenance records that fingerprint the private source files.

Prefer hand-authored, minimal synthetic fixtures. A fixture may preserve only the structural fact needed by its test:

- field names, nesting, row ordering, and value types;
- deterministic equality relationships between synthetic IDs;
- timestamp relationships needed by the scenario;
- internally consistent synthetic token arithmetic;
- the minimum synthetic text needed to classify a role, tool, or record.

Replace or omit all real prompts, outputs, reasoning, source code, diffs, tool arguments, file paths, cwd values, repository/branch/org/person names, IDs, timestamps, model response IDs, environment data, secrets, and account/provider identifiers not essential to the test.

Synthetic fixtures must be clearly labeled synthetic. Review the staged fixture diff manually and run the repository's secret/private-content checks before commit. Errors and logs must report bounded structural diagnostics, never source rows or payload text.

Cadence may be used as a schema and test-pattern reference, including its Pi parser and synthetic/approved fixtures, but do not copy private Cadence fixtures unless their license and sanitization status permit reuse. Recreate minimal Antiburn fixtures when uncertain.

## Characterization suite

Create a Pi suite parallel to `crates/antiburn-local/tests/codex_characterization.rs` and its fixture/golden layout after #243 merges.

Cover at least:

1. minimal valid v3 session;
2. user/assistant/tool-result ordering;
3. `text`, `thinking`, `toolCall`, successful tool result, and failed tool result;
4. all four usage buckets, including cache-heavy turns;
5. model attribution and a mid-session model change;
6. explicit thinking-level change;
7. compaction and known inert housekeeping rows;
8. unknown top-level row type;
9. unknown nested content-block type;
10. malformed middle record;
11. truncated final record;
12. oversized record;
13. header-only and unsupported-version sessions;
14. claimed full read, accepted prefix, and source mutation/replacement;
15. parent/fork pair plus non-fork negatives, or an explicit unsupported expectation if ownership cannot be represented;
16. streaming/batch metrics parity;
17. evidence contains no transcript text, paths, IDs, or tool arguments outside explicitly bounded safe structural labels;
18. exact capability declaration versus emitted evidence.

Use production adapter entry points in tests; do not test only detached parsing helpers.

## Non-goals

- Reworking Pi discovery or adding another Pi storage location.
- Persisting raw transcript rows, prompts, outputs, reasoning, tool arguments, paths, or source code.
- Adding pricing support for Pi's `usage.cost` object.
- Creating a new evidence schema solely for speculative Pi metadata.
- Inferring account identity, fork ownership, or subagent relationships without structural proof.
- Changing Claude or Codex behavior except for shared lists/tests required by the new provider.
- Adding telemetry or new UI surfaces.
- Promoting any other generic JSONL vendor in the same pull request.

## Acceptance criteria

- [ ] `adapter_for("pi")` resolves to a dedicated Pi adapter and `has_dedicated_adapter("pi")` is true.
- [ ] Pi metrics retain parity with existing Pi-shaped metrics tests.
- [ ] `visit` and `visit_claimed` use bounded framing and return correct source-validity outcomes.
- [ ] `SourceCapabilities::pi()` is documented, test-pinned, and exactly matches emitted evidence.
- [ ] Unknown rows/content obey the final #229 policy and surface through shared report/DTO behavior where required.
- [ ] Truncated, malformed, oversized, cancelled, and changed-source cases produce honest bounded outcomes without leaking content.
- [ ] Model transitions, tool calls/results, thinking, usage classes, and compaction are characterized.
- [ ] Fork/inherited-history ownership is either correctly implemented and tested or explicitly unsupported without double-counting.
- [ ] Pi is routed through the desktop evidence worker and added to the evidence cohort.
- [ ] Existing Pi sessions are enqueued/backfilled by the next migration and revision changes.
- [ ] Frontend/backend support flags agree and change only after the complete provider path exists.
- [ ] Pi report and badge coverage matches #242 and #229.
- [ ] Every committed fixture is minimal synthetic or independently approved public material; no private `~/.pi` content or fingerprint is present.
- [ ] Relevant provider documentation is updated.

## Verification

Use the commands documented in `CONTRIBUTING.md` and `apps/desktop/README.md`:

```sh
# Rust workspace
cargo fmt --check
cargo clippy --all-targets -- -D warnings
cargo test

# Desktop frontend
pnpm --filter @antiburn/desktop lint
pnpm --filter @antiburn/desktop type-check
pnpm --filter @antiburn/desktop test
pnpm --filter @antiburn/desktop build
```

Also run the focused Pi characterization tests and the store migration/backfill tests while iterating. Run `scripts/check-design-drift.mjs` only if styling changes unexpectedly enter scope; styling should not be needed.

## Later workflow kickoff checklist

A later implementation workflow should begin by:

1. confirming #243 is merged and recording the new baseline;
2. reading the final #229 decision and #242 coverage behavior;
3. diffing #243's adapter, desktop routing, cohort/backfill, and fork-ownership commits as the concrete template;
4. privately sampling current `~/.pi` files only under the fixture policy above;
5. writing a Pi capability matrix and synthetic scenario inventory before implementation;
6. implementing in reviewable slices: engine/characterization, desktop routing/backfill, then support-flag/report completion;
7. running privacy review on fixtures and output surfaces before the final commit;
8. using DCO sign-off (`git commit -s`) for every commit.

### PR DESCRIPTION
**Pull request by codesoda**
## Summary

- add a dedicated bounded `PiAdapter` for Pi v3 JSONL sessions
- stream metrics and evidence through the claimed-source path introduced by #243
- preserve Pi token, model, thinking, tool, compaction, and timestamp semantics
- exclude inherited fork rows using the child header timestamp without reading or retaining the private parent path
- enroll existing Pi sessions through V13 and route Pi through the desktop evidence worker, reports, badges, and support gating
- add synthetic characterization fixtures, goldens, cross-layer tests, implementation plan, support docs, and both changelogs

Closes #244.

## Stacking dependency

This draft is stacked on #243. The four rebased Codex commits are patch-equivalent to PR #243 (`git range-diff` reports `=` for all four). Merge #243 first, then refresh this branch against `main` before taking this PR out of draft.

#242 is already present in this branch through `main`. Pi characterization verifies the real merged badge policy: the supported reasoning badge can be clean, capability-missing badges remain not assessed, and partial session evidence never reads clean.

## Format and coverage decisions

- The top-level Pi row timestamp controls ordering and fork ownership.
- A child excludes rows earlier than its own session header. Missing ownership timestamps fail closed with `AttributionIncomplete`.
- The adapter checks only whether `parentSession` exists. It never reads, retains, logs, or serializes its path value.
- Pi adopts #229's recommended strict structural policy locally: nested extension payload keys are inert, while direct usage, model, role, reasoning, tool, or compaction signals fail closed.
- `customType` values are never read or persisted.
- Unknown content blocks produce bounded discriminators and at most one unusable outcome per framed record.
- Global parser, analyzer, metrics, and evidence revisions stay at `3/5/1/2`; V13 enrolls existing Pi sessions without staling Claude or Codex.

## Privacy

All committed Pi fixtures are minimal synthetic records. No real `~/.pi` transcript, prompt, response, reasoning, source code, tool argument, path, identifier, fingerprint, account value, or captured machine output is included.

Evidence retains bounded structural facts only. The source reader uses the existing 8 MiB per-record framing limit and full source-validity rechecks. Pi keeps the shared exact metrics-retention contract and adds no provider-specific total-record truncation.

## Verification

- `cargo fmt --check`, `cargo clippy --all-targets --locked -- -D warnings`, and `cargo test --locked` in `crates/antiburn-local`
  - 884 unit tests passed; one pre-existing test ignored
  - 26 Pi characterization tests passed
  - all Claude, Codex, pipeline, source-validity, memory, and doc tests passed
- `cargo fmt --check`, `cargo clippy --all-targets --locked -- -D warnings`, and `cargo test --locked` in `apps/desktop/src-tauri`
  - 575 tests passed
- `pnpm --filter @antiburn/desktop lint`
- `pnpm --filter @antiburn/desktop type-check`
- `pnpm --filter @antiburn/desktop test`
  - 88 files and 901 tests passed
- `pnpm --filter @antiburn/desktop build`
- `pnpm run slop:all`
  - score 100, zero diagnostics
- `pnpm run secrets`
- `git diff --check`
- DCO sign-off verified on every branch commit

### CODE REVIEW

File: apps/desktop/src-tauri/src/insights_report.rs [hunk 1]
```diff
@@ -543,6 +543,63 @@ mod tests {
             );
         }
 
+        #[test]
+        fn pi_backfill_moves_awaiting_support_into_the_pending_queue() {
```
Comment (daveslutzkin) [inline apps/desktop/src-tauri/src/insights_report.rs]: probably don't need this one

**Review by daveslutzkin — APPROVED** [review]

**Review by z0w0 — APPROVED** [review]
